src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -24,94 +22,6 @@ internal AsyncWindowsFileStreamStrategy(string path, FileMode mode, FileAccess a
-            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
-            // struct for this request and gets a delegate to a managed callback
-            // from there, which it then calls on a threadpool thread.  (We allocate
-            // our native OVERLAPPED structs 2 pointers too large and store EE
-            // state & a handle to a delegate there.)
-            //
-            // If, however, we've already bound this file handle to our completion port,
-            // don't try to bind it again because it will fail.  A handle can only be
-            // bound to a single completion port at a time.
-            if (handle.IsAsync != true)
-            {
-                try
-                {
-                    handle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(handle);
-                }
-                catch (Exception ex)
-                {
-                    // If you passed in a synchronous handle and told us to use
-                    // it asynchronously, throw here.
-                    throw new ArgumentException(SR.Arg_HandleNotAsync, nameof(handle), ex);
-                }
-            }
-        }
-
-        protected override void OnInit()
-        {
-            // This is necessary for async IO using IO Completion ports via our
-            // managed Threadpool API's.  This (theoretically) calls the OS's
-            // BindIoCompletionCallback method, and passes in a stub for the
-            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
-            // struct for this request and gets a delegate to a managed callback
-            // from there, which it then calls on a threadpool thread.  (We allocate
-            // our native OVERLAPPED structs 2 pointers too large and store EE state
-            // & GC handles there, one to an IAsyncResult, the other to a delegate.)
-            try
-            {
-                _fileHandle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(_fileHandle);
-            }
-            catch (ArgumentException ex)
-            {
-                throw new IOException(SR.IO_BindHandleFailed, ex);
-            }
-            finally
-            {
-                if (_fileHandle.ThreadPoolBinding == null)
-                {
-                    // We should close the handle so that the handle is not open until SafeFileHandle GC
-                    _fileHandle.Dispose();
-                }
-            }
-        }
-
-        private void TryToReuse(ValueTaskSource source)
-        {
-            source._source.Reset();
-
-            if (Interlocked.CompareExchange(ref _reusableValueTaskSource, source, null) is not null)
-            {
-                source._preallocatedOverlapped.Dispose();
-            }
-        }
-
         public override int Read(byte[] buffer, int offset, int count)
         {
             ValueTask<int> vt = ReadAsyncInternal(new Memory<byte>(buffer, offset, count), CancellationToken.None);
```

Comment:
Rather than doing all the work to create the async operation and then blocking waiting on the returned task, can this instead now use RandomAccess.ReadAtOffset, and let it handle the blocking via the overlapped?

Reply:
I would prefer not to do that. The new APIs are free of the `FileStream` backward compatibility burden and (On Windows) allow for sync operations only for sync handles. They are supposed to allow for high-performance IO, and anti-patterns are not allowed.

Reply:
> and (On Windows) allow for sync operations only for sync handles.

Where is that being checked?

Reply:
https://github.com/dotnet/runtime/blob/6241b088c5e7722613d0e641933355dbc5d69aa5/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs#L261-L268

https://github.com/dotnet/runtime/blob/6241b088c5e7722613d0e641933355dbc5d69aa5/src/libraries/System.IO.FileSystem/tests/RandomAccess/Base.cs#L84-L96

Reply:
> They are supposed to allow for high-performance IO, and anti-patterns are not allowed.

That's a good goal, and certainly for new things avoiding legacy is something to strive for.

At the same time, it feels like we might be "kicking the can down the road" as it were.  You're a developer writing a method.  It's given a SafeFileHandle and you want to use RandomAccess to do it.  Every call site then is going to have to be like, e.g. in an async method:
```C#
// in an async method
if (handle.IsAsync)
{
    await RandomAccess.ReadAsync(handle, ...);
}
else
{
    RandomAccess.Read(handle, ...);
}
```
?  And if they're writing cross-platform code, they're then doing unnecessary work when on Linux but this is required for Windows?

If we do it inside RandomAccess, we can do it as efficiently as possible.  And given that a SafeFileHandle publicly supports both modes, and the underlying support in Windows is bifurcated, this is not something I see going away any time soon.

Reply:
> You're a developer writing a method

I would rather say that you are a developer interested in high performance file IO.

Since you wanted high performance file IO and FileStream was not enough for you, you most probably know that:

* not every file is a regular file and you can't use the new APIs for sockets and pipes
* async and sync IO are two different things and should not be mixed
* blocking on async methods is bad and should be avoided for high performance scenarios

> That's a good goal, and certainly for new things avoiding legacy is something to strive for.

I would really love to at least try to achieve this goal. I really don't want to:

* increase the code complexity by copying the memory from potentially stack-allocated span to a rented array, just to support sync operations on async file handles on Windows

https://github.com/dotnet/runtime/blob/00ee1c18715723e62484c9bc8a14f517455fc3b3/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L696-L705

* block on async methods to support sync operations on async file handles

If we receive a lot of complains about the user experience, we can always stop throwing and handle the problematic scenarios like we do for `FileStream`.

Reply:
> you most probably know that

I disagree all of these are a given.  And, even if you do know all those things, you've got a SafeFileHandle and you want to read from it; you're now forcing that developer to bifurcate their code, when we could have hidden all of that for them.

> increase the code complexity

You're otherwise putting that complexity onto the caller.  And, the code is already doing the check! Rather than throwing, it can make it "just work", even if it's not 100% optimal.

> by copying the memory from potentially stack-allocated span to a rented array

This is not necessary.  An overlapped can be created that signals an event.  The code path here can then block on that event and never have to leave the stack; the pinned span is sufficient.  This also avoids the sync code path with an async handle having to block waiting for the I/O thread to invoke the callback that completes a task that is instead being blocked on.  That's why I'm pushing on this: I'm highlighting that we can make the sync-over-async case better.

Reply:
> And, even if you do know all those things, you've got a SafeFileHandle and you want to read from it; you're now forcing that developer to bifurcate their code, when we could have hidden all of that for them.

I don't want to force the developers to have two code paths. I want them to decide whether they want to use sync or async IO and based on that pass the right arguments to `File.OpenHandle` and then use sync or async methods.
This is why I wanted to introduce two primitive types in the first place (one for async and one for sync IO).

> You're otherwise putting that complexity onto the caller.

This API was designed for high-performance, not maximum usability. If someone does not like it, they can keep using `FileStream`.

> I'm highlighting that we can make the sync-over-async case better.

I agree, but I don't want to increase the code complexity on our side. We haven't received any complaints about the performance of sync methods of `FileStream` opened for async IO.

Reply:
> I want them to decide whether they want to use sync or async IO and based on that pass the right arguments to File.OpenHandle and then use sync or async methods.

This assumes they own and control all such entire code paths, from handle open through all usage.  And it assumes all code paths on that same handle are able to be either all sync or all async.  That's typically not how things play out in real apps, especially ones that aren't entirely greenfield.  It's also forcing any libraries that want to accept a SafeFileHandle to either propagate this same philosophy and kick the can down the road or implement two code paths.

> This API was designed for high-performance, not maximum usability

Making this more usable detracts nothing from the performance.  It doesn't even add any more branches.  It's just turning something that's currently exceptional into something that works.

> We haven't received any complaints about the performance of sync methods of FileStream opened for async IO.

We have.  But even if we hadn't, we make things better all the time even if specific complaints haven't made their way to us.  That's actually the majority of performance improvements we make.  Most developers don't spend as much time profiling and analyzing as we do, and often don't know what's consuming time in their applications, or even if they do, know whether it could be better or how much better it could be.

Reply:
We don't need to block this PR to address this.  But I think it'd be valuable to track it.  From my perspective, it's a noticeable gap.

Reply:
> We don't need to block this PR to address this. But I think it'd be valuable to track it. From my perspective, it's a noticeable gap.

Reply:
> We don't need to block this PR to address this. But I think it'd be valuable to track it. From my perspective, it's a noticeable gap.

Is there any chance that you could create an issue for that and provide some hints about the implementation? Mostly about this part:

> An overlapped can be created that signals an event. The code path here can then block on that event and never have to leave the stack; the pinned span is sufficient. This also avoids the sync code path with an async handle having to block waiting for the I/O thread to invoke the callback that completes a task that is instead being blocked on.

Reply:
> provide some hints about the implementation? Mostly about this part

Just threw this together so take it all with grains of salt, but e.g.
```C#
#pragma warning disable CA1416
using Microsoft.Win32.SafeHandles;
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;

unsafe class Program
{
    static void Main()
    {
        string path = Path.GetTempFileName();
        File.WriteAllText(path, string.Concat(Enumerable.Range(0, 10_000).Select(i => (char)('a' + (i % 26)))));

        using var sfh = CreateFile(path, FileAccess.Read, FileShare.Read, IntPtr.Zero, FileMode.Open, (int)FileOptions.Asynchronous, IntPtr.Zero);
        using var mres = new ManualResetEvent(false);

        NativeOverlapped overlapped = default;
        overlapped.EventHandle = mres.SafeWaitHandle.DangerousGetHandle();

        Span<byte> onStackBuffer = stackalloc byte[1024];
        fixed (byte* p = &MemoryMarshal.GetReference(onStackBuffer))
        {
            int r = ReadFile(sfh, p, onStackBuffer.Length, IntPtr.Zero, &overlapped);
            Debug.Assert(r == 0);
            Debug.Assert(Marshal.GetLastPInvokeError() == 997); // ERROR_IO_PENDING
        }
        mres.WaitOne();

        int numBytes;
        Console.WriteLine(GetOverlappedResult(sfh, &overlapped, &numBytes, false));
        Console.WriteLine(numBytes);
        Console.WriteLine(Encoding.ASCII.GetString(onStackBuffer.Slice(0, numBytes)));
    }

    [DllImport("kernel32", EntryPoint = "CreateFileW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false, ExactSpelling = true)]
    private static extern unsafe SafeFileHandle CreateFile(
        string lpFileName,
        FileAccess dwDesiredAccess,
        FileShare dwShareMode,
        IntPtr lpSecurityAttributes,
        FileMode dwCreationDisposition,
        int dwFlagsAndAttributes,
        IntPtr hTemplateFile);

    [DllImport("kernel32", SetLastError = true)]
    private static extern unsafe int ReadFile(SafeHandle hFile, byte* lpBuffer, int nNumberOfBytesToRead, IntPtr lpNumberOfBytesRead, NativeOverlapped* lpOverlapped);

    [DllImport("kernel32", SetLastError = true)]
    private static extern unsafe bool GetOverlappedResult(SafeFileHandle hFile, NativeOverlapped* lpOverlapped, int* lpNumberOfBytesTransferred, bool bWait);
}
```

Reply:
@stephentoub I've given it a try and failed to make "mixed" (sync and async IO using same handle) mode working. The problematic scenario:

* open handle for async IO
* perform async IO (it initializes `ThreadPoolBinding`)
* try to perform sync IO using the same handle.

I end up getting:

```log
Unhandled exception. Fatal error. Internal CLR error. (0x80131506)
```

It's not easy to debug and from my initial observations it seems that once we initialize `ThreadPoolBinding` we always receive callback no matter if we specify the `EventHandle` or not.


The code can be found here: https://github.com/dotnet/runtime/commit/abffffdcfa1cf2ac7be705ec66cc18d5cc77e4f9

@stephentoub do you have any ideas for how to fix it? 

It would be great if you could approve the PR so I could merge it today or tomorrow before we snap for Preview 6.

Reply:
> do you have any ideas for how to fix it?

If the handle is already bound to the thread pool, then the implementation would need to ensure the NativeOverlapped was allocated rather than be on the stack, and the callback would release it.  It's more complicated, but still possible.

In the meantime, making the API "just work" by delegating to the sync or async API is straightforward and is what we do in .NET.

Reply:
Thanks for a great hint, I got it working and will  send a separate PR tomorrow